### ISSUE TITLE
Branch naming and Gitflow oversight

### ISSUE THREAD

**Issue by buscandoaverroes** [worldbank/gld#146]
This will take a while to walk through, so my apologies. But in short, I think it might be beneficial to allow for unique country develop-branch names in the [naming schema](https://worldbank.github.io/gld/docs/branch_management/bm_index/#naming-rules-for-branches). Ex: `MEX-1`, `ZAF-7`, etc.

Let me explain why through links, pictures, and Gitflow.

The general structure of our workflow is parallel, but it's not entirely isolated: code updates from one develop branch are often needed in another. Let's take the case of this recent issue, #142 , ISCO as String. In this instance, let's say we have the following branches:

- develop-merge
- IDN
- IND
- MEX
- PHL
- ZAF
- TUR

I'm just going to pick the first two for simplicity but the case applies to all of them. Here's a parallel snapshot of IDN and develop-merge. 

![issue1](https://user-images.githubusercontent.com/56852146/133627580-fbd234a8-b2dc-4c2a-a7ef-55bfd534a9f8.jpg)

The star ( ⭐ ) commit in develop-merge indicates an update that IDN needs -- similar to [naming schema](https://worldbank.github.io/gld/docs/branch_management/bm_index/#naming-rules-for-branches) when @gronert-m updated the code for GLD checks that we all would need. How should that updated in (⭐ ) go to IDN?

From [this guide](https://www.atlassian.com/git/tutorials/merging-vs-rebasing) there are two principle options in `git`:

1. Merge the branches, or 
2. Rebase IDN from develop-merge

**Merging**
Merging is 100% what we've been doing so far, via a pull request. From what I understand, merging is generally preferred because it preserves the commit history of both branches. Of course, when you merge via pull request, you update the branch you merge into, so, in this case, develop-merge gets all the updates from IDN in order for the harmonizer in IDN to get the (⭐) update.

![issue2](https://user-images.githubusercontent.com/56852146/133629695-ad2b57ce-c396-45ed-8c57-1bbb4ada03f9.jpg)

It's considered best practice to delete the IDN branch after you merge from the remote (cloud) to avoid clutter, and because after the merge, having IDN becomes redundant and necessary: all commits in IDN are in develop-merge. There's no more IDN now.

However, `git` keeps a copy of each branch on the remote and on your local computer: when the harmonizer deletes IDN on the remote, *a copy stays locally on the harmonizer's computer by default* (I believe). I think this is so you can restore the branch later if problems occur, and probably other reasons. This remote-local dynamic is important because....

You can't really do this:

![issue3](https://user-images.githubusercontent.com/56852146/133633018-496d3a08-258f-4887-babf-714950e1709f.jpg)

*The harmonizer cannot recreate IDN after merging with develop-merge because this branch already exists elsewhere in the repository history* (i.e., on the harmonizer's local computer). If we follow the naming guidelines [as they stand now ](https://worldbank.github.io/gld/docs/branch_management/bm_index/#naming-rules-for-branches) the new IDN develop branch should be called `IDN`. I see a few ways to get around this, but stating my opinion for number 2:

1. Do not modify the guidelines and begin a `git rebase` workflow
2. Modify the naming guidelines to allow for numbers or some quasi-infinite iteration to develop branches

The reason I suggest 2 is because `git rebase` can be risky if you're not careful. Let me explain. 

**Rebase**
Instead of merging the changing of IDN into develop-merge, we could `git rebase` (see the guide above for the terminal command). Here's what this does:

- leaves develop-merge untouched
- takes all changes from develop-merge and adds commits from IDN "after" those

![issue4](https://user-images.githubusercontent.com/56852146/133632595-a2c006ad-1a81-498e-9641-ad74dcddce80.jpg)

How is this actually accomplished? You can do this in [GitHub Desktop](https://docs.github.com/en/desktop/contributing-and-collaborating-using-github-desktop/keeping-your-local-repository-in-sync-with-github/syncing-your-branch#rebasing-your-project-branch-onto-another-branch) or in [GitKraken](https://www.gitkraken.com/learn/git/problems/git-interactive-rebase). However, I think I would prefer doing this in GitKraken because you can perform an "interactive rebase", which essentially gives you more control, and I don't think you can do this in GH desktop.

If we really really cared about keeping our naming scheme the way it is, this may solve our issue. But we would have to rebase *every* time to get changes in another branch like (⭐), which we've seen happens more often than we think. Also, in terms of repo cleanliness, it sort of messes with branch histories: when you rebase IDN, you are re-writing the commit history of IDN. I personally don't like that, but there are competing organizational theories out there.

In summary, I think the least-destructive option is to create some sort of counter that allows for semi-infinite iterations of develop-branches. 

- [AAA]-#
- [AAA]#
- [AAA]-[random 3 digit string]

This way, we can merge or rebase as desired, and the branch can be merged and re-created as many times as needed. It is slightly annoying, but I think we will all run into the duplicate branch problem eventually. Thoughts and opinions are obviously welcome here and I think we should obviously discuss this at our next meeting.

**Comment by gronert-m** [comment]
Thanks @buscandoaverroes for highlighting this. I have read the article from Bitbucket. I think one line that is important here was what is said about constant remerging:

>  If main is very active, this can pollute your feature branch’s history quite a bit. While it’s possible to mitigate this issue with advanced git log options, it can make it hard for other developers to understand the history of the project.

I guess the question here is:

* Do we update main often in a way that requires feature branches to have those updates?
*  Do we worry about polluting the feature branch's history?

In the first case, it is not when a new harmonization is added to `main` that is relevant (adding `PHL` has no impact on feature `IDN`) but the general "support" kind of things. This has been happening of late though I am unsure how much of it is regular upkeep, how much is growing pains. If it is the former (and I read you as seeing it that way) it is necessary, if it is the latter, less so. I wish it were the latter, but am uncertain now.

On the second point, I think polluting the feature branch's history is not a problem, as it is thought for eventual deletion anyway.

My feeling so far is: rebasing is a good idea, but I would leave it up to whoever is running the feature branch.

Then there is the second part about names not being able to be repeated. This I did not understand. I just created a new branch called `prueba` from `main`. I made a small amendment to the readme file and committed it, so `prueba` was one ahead of `main`. Then I deleted `prueba` and created a new branch called `prueba`. This was not the old `prueba` (i.e., it was equal to `main` and not one ahead) and at no point did the system tell me the name was blocked.

I did this exclusively on the website via browser. I would request you run such a scenario with a branch you create locally on your computer and push to the web, then delete and recreate to confirm that what I did in the web is also possible via local copies on the hard drive. If deleting and recycling names is possible, would this not make the [CCC]-[###] naming convention redundant? (Which is independent of the rebasing debate).
